### PR TITLE
Allow to rename nodes with IDs, will fix #3974 and #4413

### LINKED ISSUE #3974: Conflicts appears when changing node_name on agents
### Description of the Issue (and unexpected/desired result)

- When changing node_name of an agent, we observe conflicts in consul servers logs.
- It is also possible (not reproduced yet, but observed twice on our production) that consul servers ends up blocked when several agents have changed their names.

### Reproduction steps

- change node_name configuration in a consul agent configuration
- restart consul agent

On the consul server:
```
2018/03/20 12:53:56 [INFO] serf: EventMemberJoin: consul-relay01-test2-pa4.central.criteo.preprod 10.224.45.123
2018/03/20 12:53:56 [INFO] consul: member 'consul-relay01-test2-pa4.central.criteo.preprod' joined, marking health alive
2018/03/20 12:53:56 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: node ID "4d1dac9f-8977-4258-4aca-fa254c9f48da" for node "consul-relay01-test2-pa4.central.criteo.preprod" aliases existing node "consul-relay01-pa4.central.criteo.preprod"
2018/03/20 12:54:00 [WARN] consul.fsm: EnsureRegistration failed: failed inserting node: node ID "4d1dac9f-8977-4258-4aca-fa254c9f48da" for node "consul-relay01-test2-pa4.central.criteo.preprod" aliases existing node "consul-relay01-pa4.central.criteo.preprod"
```

Output of consul members:
```
consul-relay01-pa4.central.criteo.preprod                10.224.45.123:8301  failed  client  1.0.6  2         pa4  <default>
consul-relay01-test2-pa4.central.criteo.preprod          10.224.45.123:8301  alive   client  1.0.6  2         pa4  <default>
```

### `consul version` for both Client and Server
Client: `1.0.6`
Server: `1.0.6` (with some patches)

### `consul info` for both Client and Server
Client:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 3
	services = 4
build:
	prerelease = 
	revision = 
	version = 1.0.6
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 32
	goroutines = 46
	max_procs = 2
	os = linux
	version = go1.9.4
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 3715
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1133006
	members = 453
	query_queue = 0
	query_time = 1929

```
Server:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 2
	services = 3
build:
	prerelease = criteo5
	revision = 
	version = 1.0.6
consul:
	bootstrap = false
	known_datacenters = 2
	leader = true
	leader_addr = 10.224.47.92:8300
	server = true
raft:
	applied_index = 83407159
	commit_index = 83407159
	fsm_pending = 0
	last_contact = 0
	last_log_index = 83407159
	last_log_term = 98
	last_snapshot_index = 83403466
	last_snapshot_term = 98
	latest_configuration = [{Suffrage:Voter ID:4fd4772d-e3cd-ebd6-731f-c7e6431ce284 Address:10.224.46.86:8300} {Suffrage:Voter ID:1bfb896b-ee04-520c-10e0-8b382dc0c832 Address:10.224.47.92:8300} {Suffrage:Voter ID:f5e7d35e-66b4-a4b1-b85b-5747af533b58 Address:10.224.47.83:8300}]
	latest_configuration_index = 77270117
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 98
runtime:
	arch = amd64
	cpu_count = 32
	goroutines = 2592
	max_procs = 31
	os = linux
	version = go1.10
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 3715
	failed = 2
	health_score = 0
	intent_queue = 0
	left = 1
	member_time = 1133006
	members = 456
	query_queue = 0
	query_time = 1929
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 6511
	members = 6
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details

centos7.3

### PR DESCRIPTION
**Pull request by pierresouchay**
This change allow to rename any well behaving recent agent with an
ID to be renamed safely, ie: without taking the name of another one
with case insensitive comparison.

Deprecated behaviour warning
----------------------------

Due to asceding compatibility, it is still possible however to
"take" the name of another name by not providing any ID.

Note that when not providing any ID, it is possible to have 2 nodes
having similar names with case differences, ie: myNode and mynode
which might lead to DB corruption on Consul server side and
lead to server not properly restarting.

See #3983 and #4399 for Context about this change.

Disabling registration of nodes without IDs as specified in #4414
should probably be the way to go eventually.

### CODE REVIEW

**Comment by pierresouchay** [comment]
@banks @mkeeler Ok, I implemented exactly what you said in https://github.com/hashicorp/consul/pull/4399#issuecomment-406056900

The only differences are (apart from being a single commit):
 - another test when performing a registration with an ID (see https://github.com/hashicorp/consul/pull/4415/files#diff-2271ca70cf81490eff6dd6060d3f35b5R396 )
 - more tests, with explicit comments about deprecated behavior

Hope you will be satisfied with this

Kind regards

**Comment by pierresouchay** [comment]
@mkeeler I had to remove the else test for case insensitive search in my second commit since it broke test `TestAgentAntiEntropy_Services`. While that's probably fine for renaming, it means it won't fix properly https://github.com/hashicorp/consul/issues/4413 :(

Still some room for further PRs then

**Comment by pierresouchay** [comment]
@mkeeler I think I got it right now. I just added `&& enode.ID != "" ` to avoid collisions with nodes without IDs and all tests do pass.

So, it basically fixes https://github.com/hashicorp/consul/issues/4413 and https://github.com/hashicorp/consul/issues/3974 again !

File: agent/consul/state/catalog.go [hunk 1]
```diff
@@ -365,21 +380,44 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
...
+					return fmt.Errorf("Error while renaming Node ID: %q: %s", node.ID, dupNameError)
+				}
+				// We are actually renaming a node, remove its reference first
+				err := s.deleteNodeTxn(tx, idx, n.Node)
+				if err != nil {
+					return fmt.Errorf("Error while renaming Node ID: %q from %s to %s",
+						node.ID, n.Node, node.Node)
+				}
+			}
+		} else {
```
Comment (mkeeler) [inline agent/consul/state/catalog.go]: The else statement here is one of the problematic ones. I would recommend replacing with this statement.

```
}
// TODO: else statement missing to ensure name is not already taken by another node
// If added right now it could break existing behavior which allows for adding
// a UUID to a node that doesn't have one when the name is the same.
```

Comment (pierresouchay) [reply agent/consul/state/catalog.go]: @mkeeler I think the else statement is now SAFE since `s.ensureNoNodeWithSimilarNameTxn` will not match any node without an ID. It means it will break ONLY if: a node having an ID clashes (case insensitive) with another node having an ID.

In all other cases, the behaviour is unchanged as shown in tests `deprecatedEnsureNodeWithoutIDCanRegister` of agent/consul/state/catalog_test.go.

# What works:

## When request has an ID:
* renaming a node when having an ID, without colliding with the name of another node (case insensitive)
* creating a node with an ID for which the name is not taken by anybody (case insensitive)

## When request has not ID (more permissive regarding case insensitivity)
* Create a new node without ID with name not colliding
* Remove a node ID of a node with an ID if name EXACTLY (case sensitive) matches
* Create a new node without ID with name colliding with case insensitive differences

# What does not work anymore due to this change:
* Create a node with an ID, but this name collides with another one (case insensitive) having an ID (works if other node DOES not have any ID)

# Previously, given the following configuration

id42 - node1

Registering same name with different ID is blocked
id43 - node2 => previously would 'steal' node2 to id42    ==> now blocked
id43 - NoDe2 => previously would create a second node (id42: node2 ; id43: NoDe2) => the big crash I had described here https://github.com/hashicorp/consul/issues/4413 => now the registration is blocked and id43 won't be able to join

# Conclusion

* A request with an ID can "steal" a node without any ID, but cannot steal a node with an ID if ID is different
* A request without any ID can do whatever it wants

So, basically, the ONLY new possible issue we have is:
 - if you change manually the ID of a node, this node won't be able to join the cluster
 - if 2 nodes have different IDs and case insensitive differences, only ONE can join, which is good new for DNS as well as it avoid the DB corruption I had.

What do you think of this ?

Comment (mkeeler) [reply agent/consul/state/catalog.go]: I think I understand now.

We will error out if both if an existing node with a uuid != new uuid where there is a name match but we will still allow adding a uuid to a node that doesn't have one. You just moved the logic to only enforce it when they both have uuids into that function.

File: agent/consul/state/catalog.go [hunk 2]
```diff
@@ -365,21 +380,44 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
 		if existing != nil {
 			n = existing.(*structs.Node)
 			if n.Node != node.Node {
-				return fmt.Errorf("node ID %q for node %q aliases existing node %q",
-					node.ID, node.Node, n.Node)
+				// Lets first get all nodes and check whether name do match
+				dupNameError := s.ensureNoNodeWithSimilarNameTxn(tx, node)
```
Comment (mkeeler) [inline agent/consul/state/catalog.go]: I think there is a bug here.

Because the name check only will match if the other node has a UUID and the UUID doesn't match as well, this will potentially allow merging two existing nodes into 1.

An example of this is if there are two existing nodes: node foo has a uuid and node bar does not. If you attempt to rename foo to bar it will not error out for the name match since the bar node has no uuid. It will delete node foo and then add the uuid to the bar node and update all its info. This is definitely not desired.

Adding a uuid to a node (the else case where we had no uuid match) is okay to allow for upgrading to uuids. However we need to prevent this case where we have two existing nodes and we merge them into one.

Comment (pierresouchay) [reply agent/consul/state/catalog.go]: @mkeeler Well spotted, I introduced this bug after I saw `TestAgentAntiEntropy_Services` not working  => should be fixed in next patch

Renaming is not allowed when another node, including without nodeID exists in next patch.

**Comment by pierresouchay** [comment]
@mkeeler Something is wrong... I saw by adding some tests for renaming. It seems that index ("nodes", "uuid") is not properly updated... I think indexing using "uuid" for nodes does not work well.

I'll keep you updated

**Comment by pierresouchay** [comment]
@mkeeler The last patch fix an existing bug: GetNodeID() and similar lookups were not working properly since index uuid was UUIDIndex while lookups were performed on strings. Thus, we were never entering in the test for renaming.

This fixes both and include tests cases that includes all we discussed, this time should be the good one ! :-)

**Comment by pierresouchay** [comment]
@mkeeler the last patch explains why I never got the "aliases existing node" in prod when renaming nodes, basically, the feature was not working at all since lookup per node uuid did not work at all, what do you think about the change now?

**Comment by pierresouchay** [comment]
Hello @banks & @mkeeler

In my last patches, I just improved the error messages when `GetNodeId()` get a wrong UUID and improved the test coverage for `GetNodeId()` function

So basically this review:
* `"node ID %q for node %q aliases existing node %q"` => was never working due to wrong lookup by UUID since `GetNodeID()` nor the test for entering this condition was working=> always in the case of renaming anyway (and not protected against case insensitivity)
* `GetNodeID()` is now working as expected (was not working due to indexation working on UUID binary representation while lookup performed on string
* Changed the behavior as explained above: https://github.com/hashicorp/consul/pull/4415#discussion_r203755303

I will wait for a review anyway now, I think we are good

File: agent/consul/state/catalog.go [hunk 3]
```diff
@@ -350,6 +351,23 @@ func (s *Store) EnsureNode(idx uint64, node *structs.Node) error {
...
+func (s *Store) ensureNoNodeWithSimilarNameTxn(tx *memdb.Txn, node *structs.Node, allowClashWithoutID bool) error {
+	// Retrieve all of the nodes
+	enodes, err := tx.Get("nodes", "id")
+	if err != nil {
+		return fmt.Errorf("Cannot lookup all nodes: %s", err)
+	}
+	for nodeIt := enodes.Next(); nodeIt != nil; nodeIt = enodes.Next() {
+		enode := nodeIt.(*structs.Node)
+		if strings.EqualFold(node.Node, enode.Node) && node.ID != enode.ID {
+			if !(enode.ID == "" && allowClashWithoutID) {
```
Comment (banks) [inline agent/consul/state/catalog.go]: Does this cover the case where the _new_ node (node.ID == "") but clashing node isn't? That is specifically what broke tests before.

I'll read on and may well find a reason why that can't happen here but just noting. Even if that is a true invariant might be worth commenting because we want to allow either the existing or the new node to have blank ID without changing behaviour.

File: agent/consul/state/catalog.go [hunk 4]
```diff
@@ -358,28 +376,51 @@ func (s *Store) ensureNodeTxn(tx *memdb.Txn, idx uint64, node *structs.Node) err
...
+			}
+		} else {
+			// We are adding a node with an ID, ensure name is not already taken by another node
+			dupNameError := s.ensureNoNodeWithSimilarNameTxn(tx, node, true)
+			if dupNameError != nil {
+				return fmt.Errorf("Error while renaming Node ID: %q: %s", node.ID, dupNameError)
 			}
 		}
+		// TODO: a else statement is missing here to check we are not stealing
+		// a similar case-insensitive name
```
Comment (banks) [inline agent/consul/state/catalog.go]: Is this comment still accurate? I thought that is what this PR is meant to fix?

File: agent/consul/state/catalog_test.go [hunk 5]
```diff
@@ -401,6 +453,189 @@ func TestStateStore_EnsureRegistration_Restore(t *testing.T) {
...
+	}
+
+	// Node has updated information, its ID has been removed (deprecated, but working)
+	if idx != 7 || out.Node != firstNodeName || out.ID != "" || out.Address != "1.1.7.10" {
+		t.Fatalf("[DEPRECATED] bad node returned: %#v", out)
+	}
+	if out.CreateIndex != 1 || out.ModifyIndex != 7 {
+		t.Fatalf("[DEPRECATED] bad CreateIndex/ModifyIndex returned: %#v", out)
+	}
+}
```
Comment (banks) [inline agent/consul/state/catalog_test.go]: There is no test for the reverse case which is behaviour our test suite (and possibly other clients) rely on:

 - A node is registered _with_ an ID (by a real agent starting)
 - Later tooling re-registers the same node (say to update it's node meta) but doesn't bother to specify the `ID` field (because it never had to before)

The above should continue to work and replace the existing node with the new one _without an id_.

All that said, CI shows that the same tests are not broken _right now_ but they were only reliably broken before due to another change we reverted. They could still be non-deterministically broken since timing was involved.

Specifically in tests like `TestPrepaparedQuery_Execute`:
 - The test agent starts up assigns itself a UUID and _in the background_ Anti-entroypy _eventually_ registers that with the server (normally in milliseconds but that is enough of a delay to make it non-deterministic...)
 - Then actual test code like this: https://github.com/hashicorp/consul/blob/fa0d8aff335a5ea169b05cd2436358d2a77a07dc/agent/consul/prepared_query_endpoint_test.go#L1992
 - Re-registers the same node name to change it's health status. This registration happens not to have an ID so any of the following orderings can happen:
    - Agent AE gets there first so the _existing_ node in state _does_ have an ID while the incoming registration _doesn't_  - this is the case that is missing an explicit test here I think
    - test code registers first with no ID and then the AE upgrades it to have an ID. Then subsequent code in `setHealth` registers again without an ID.

It's pretty clear that the test is racey anyway regardless of the ID thing, but this is why we think it's important that we don't break the case where and existing node DOES have an ID but a registration doesn't specify one. We can always fix this test and arguably should, but other people might have legitimately built something that does something similar to this test and we can't break them.

Is that clear?

Now it _might_ be that your code is correct already and will always pass that test regardless of timing/interleaving - esp. given that it's not failing now. but I don't see an explicit test case here that verifies this behaviour and the logic is complicated enough I'm not sure I trust my self to reason about if it works alone 😄.

Could we add that?

File: agent/consul/state/catalog.go (hunk 3 above)
Comment (pierresouchay) [reply agent/consul/state/catalog.go]: This function does not, but basically yes since this function is called ONLY when node.ID != ""

In my next and (hopefuly) last patch, I did not change anything but added tests to this function.

# To summarize

## When ID is provided:
 - if the node already exists, you rename the node if no conflicts with another node
 - if node does not exists, check for conflicts, but ONLY for nodes having ID (thus we allow to "upgrade" a node without ID and add an ID)

## When ID is not provided:
 - free lunch, you can "downgrade" a node by removing its ID - BUT ONLY if you have the exact same name as an existing node without ID (Kept for your tests and compatibility reasons)
 - you can create a node of course

So, it means this PR is fixing #4413 (not being able to have two different agents with with that differs only by their case) ONLY in clusters where everybody is playing nice and has an ID (which is your case in our clusters :-)

As long as you have agents that register without any ID, you are in danger, you may have `node2` and `NoDe2`. But we keep all of this for the ascending compatibility with old clusters as well as your tests

File: agent/consul/state/catalog.go (hunk 4 above)
Comment (pierresouchay) [reply agent/consul/state/catalog.go]: I am removing it in next patch.

File: agent/consul/state/catalog_test.go [hunk 6]
```diff
@@ -73,6 +73,53 @@ func TestStateStore_GetNodeID(t *testing.T) {
...
+	if err := s.ensureNoNodeWithSimilarNameTxn(tx, node, true); err == nil {
+		t.Fatalf("Should return an error since another name with similar name exists")
+	}
+	// Lets conflict with node without ID
+	node.Node = "NoDe2"
+	if err := s.ensureNoNodeWithSimilarNameTxn(tx, node, false); err == nil {
+		t.Fatalf("Should return an error since another name with similar name exists")
+	}
+	if err := s.ensureNoNodeWithSimilarNameTxn(tx, node, true); err != nil {
+		t.Fatalf("Should return an error since another name with similar name exists")
```
Comment (banks) [inline agent/consul/state/catalog_test.go]: Heh I was about to write a whole other thing about how this still doesn't exercise the no-ID free lunch case we've been talking about but then I looked closer and spotted this final case has `err != nil` 😄 .

The message here though still asserts the _opposite_ of the condition though which is what confused me - it should be "Should not return an error since re-registering an identical node name without an ID should not be broken" or something.

Other than that I think this is good now.

Comment (pierresouchay) [reply agent/consul/state/catalog_test.go]: @banks Thank you, fixed the error message in test case in next patch!

**Comment by pierresouchay** [comment]
@mkeeler Do you feel comfortable with the PR now?

Kind Regards

**Comment by pierresouchay** [comment]
@mkeeler Is the review Ok now?

We had 3 incidents (including one major one) to to the "renaming", this PR is really important to us.

**Comment by mkeeler** [comment]
Sorry about that @pierresouchay . Apparently I had done everything besides hit the "Submit Review" button on Monday

File: agent/consul/state/catalog.go [hunk 7]
```diff
@@ -421,6 +463,32 @@ func (s *Store) GetNode(id string) (uint64, *structs.Node, error) {
 	return idx, nil, nil
 }
 
+func getNodeIDTxn(tx *memdb.Txn, id types.NodeID) (*structs.Node, error) {
+	strnode := string(id)
```
[the remark this answers is not in the document]
Comment (pierresouchay) [reply agent/consul/state/catalog.go]: DONE

**Comment by pierresouchay** [comment]
> Sorry about that @pierresouchay . Apparently I had done everything besides hit the "Submit Review" button on Monday

@mkeeler No problem :-) The fix is DONE regarding the parsing of uuid anyway.

Have a great Week-End or Holidays

**Comment by mkeeler** [comment]
Before merging this I am going to pull down the branch and make sure all unit tests run as expected. Travis seems to always have different errors each time.

**Comment by pierresouchay** [comment]
@mkeeler Sure... I am trying to fix Travis, but it takes time: https://github.com/hashicorp/consul/pull/4472 :( almost there

**Comment by pierresouchay** [comment]
@mkeeler Since #4472 has been merged, do you want me to rebase so we can see whether the tests do pass now?

**Comment by pierresouchay** [comment]
@mkeeler any news on your tests?
That's currently the Consul bug that causes the most incidents in our clusters, we really would like to include this fix in next releases, that's would be great

**Comment by mkeeler** [comment]
@pierresouchay All tests pass locally.

**Comment by pierresouchay** [comment]
@mkeeler \o/ Thank you so much
